module/VuFind/src/VuFind/View/Helper/Root/ProxyUrl.php

```diff
@@ -66,10 +80,66 @@ public function __construct($config = null)
      */
     public function __invoke($url)
     {
-        $usePrefix = !isset($this->config->EZproxy->prefixLinks)
-            || $this->config->EZproxy->prefixLinks;
+        $useWebService = $this->config->EZproxy->prefixLinksWebServiceUrl ?? false;
+        if ($useWebService) {
+            $usePrefix = $this->checkUrl($url) ?? $this->checkConfig();
+        } else {
+            $usePrefix = $this->checkConfig();
+        }
+
         return ($usePrefix && isset($this->config->EZproxy->host))
             ? $this->config->EZproxy->host . '/login?qurl=' . urlencode($url)
             : $url;
     }
+
+    /**
+     * Return the configured prefixLinks setting.
+     *
+     * @return bool The configured setting, or the default
+     */
+    protected function checkConfig()
+    {
+        return $this->config->EZproxy->prefixLinks ?? true;
+    }
+
+    /**
+     * Check whether the given URL requires the proxy prefix.  Cache the response.
+     *
+     * @param string $url The raw URL to check
+     *
+     * @return mixed Whether the URL should be prefixed, or null if it can't be determined
+     */
+    protected function checkUrl($url)
+    {
+        $domain = parse_url($url, PHP_URL_SCHEME)
+         . '://'
+         . parse_url($url, PHP_URL_HOST);
+        $cacheKey = "proxyUrl-domainToUsePrefix-$domain";
+        $usePrefix = $this->getCachedData($cacheKey);
+        if (null === $usePrefix) {
+            $usePrefix = $this->queryWebService($domain);
+            $this->putCachedData($cacheKey, $usePrefix);
+        }
+        return $usePrefix;
+    }
+
+    /**
+     * Query the web service on whether to prefix URLs to a given domain.
+     *
+     * @param $domain The domain
+     *
+     * @return mixed Whether the URL should be prefixed, or null if it can't be determined
+     */
+    protected function queryWebService($domain)
+    {
+        $prefixLinksWebServiceUrl = $this->config->EZproxy->prefixLinksWebServiceUrl;
+        try {
+            $response = $this->httpService->get($prefixLinksWebServiceUrl, ['url' => $domain]);
+            $responseData = $response->getContent();
+        } catch (Exception $ex) {
+            $this->logError('Exception during EZproxy web service request: ' . $ex->getMessage());
+            return null;
+        }
+        return '1' === $responseData;
```

Comment:
Do you think it might be a good idea to trim $responseData? I tested this by putting a text file in VuFind's public directory and pointing the configuration at that -- I could then edit the file to change it to 0 or 1 to test both cases. However, my editor initially added line breaks, and `1\n` is interpreted as "not 1." This caused me a little bit of confusion, and I imagine it's possible if somebody builds their own custom web service that the responses might end up having stray whitespace in them.

Reply:
Agreed, done.